src/main/java/com/walmartlabs/x12/util/split/X12TransactionSplitter.java

```diff
@@ -0,0 +1,338 @@
+            transactionHolders.addAll(transactionsInGroup);
+            
+            // what's next
+            // could be a GS or IEA
+            if (segments.hasNext()) {
+                currentSegment = segments.next();
+                currentSegmentId = currentSegment.getIdentifier();
+                if (StandardX12Parser.ISA_TRAILER_ID.equals(currentSegmentId)) {
+                    // end of the envelope
+                    break;
+                } else {
+                    // not the end of envelope
+                    // assume it starts another group
+                    // so move back and process
+                    segments.previous();
+                }
+            } else {
+                this.throwParserExceptionUnexpectedEnd(currentSegmentId);
+            }
+        }
+
+        this.populateIsaEnvelope(isaHeader, iseTrailer, transactionHolders);
+        
+        return transactionHolders;
+    }
+    
+    private void populateIsaEnvelope(X12Segment isaHeader, X12Segment iseTrailer,
+        List<TransactionHolder> transactionHolders) {
+
+        transactionHolders.forEach(txHolder -> {
+            txHolder.setIsaHeader(isaHeader);
+            txHolder.setIseTrailer(iseTrailer);
+        });
+    }
+    
+    private List<TransactionHolder> doGroup(SegmentIterator segments) {
+        List<TransactionHolder> transactionsInGroup = new ArrayList<>();
+        
+        X12Segment groupHeader = null;
+        
+        X12Segment currentSegment = segments.next();
+        String currentSegmentId = currentSegment.getIdentifier();
+        
+        //
+        // first segment better be GS (start of group)
+        //
+        if (StandardX12Parser.GROUP_HEADER_ID.equals(currentSegmentId)) {
+            groupHeader = currentSegment;
+            
+            // loop until we find the GE (end of group)
+            while (segments.hasNext()) {
+                TransactionHolder transactionHolder = this.doTransaction(segments);
+                transactionHolder.setGsHeader(groupHeader);
+                transactionsInGroup.add(transactionHolder);
+                
+                // what's next
+                // could be an ST or GE
+                if (segments.hasNext()) {
+                    currentSegment = segments.next();
+                    currentSegmentId = currentSegment.getIdentifier();
+                    if (StandardX12Parser.GROUP_TRAILER_ID.equals(currentSegmentId)) {
+                        // end of the group
+                        
+                        X12Segment groupTrailer = currentSegment;
+                        
+                        // add group trailer to all transactions
+                        transactionsInGroup.forEach(txHolder -> {
+                            txHolder.setGeTrailer(groupTrailer);
+                        });
+                        
+                        break;
+                    } else {
```

Comment:
not sure follow your thought
the current logic
- after finishing a transaction ST/ST see what we get
--  if nothing is left it is an error b/c should at least have a GE (and then IEA)
-- if GE then group end was found so we should break out of doGroup method
-- otherwise assume it is an ST and let the loop try to process the transaction (doTransaction)
--- this method requires the "next" segment to be an ST